apisix/plugins/traffic-split.lua

```diff
@@ -0,0 +1,294 @@
+        end
+    end
+
+    return node
+end
+
+
+local function set_upstream(upstream_info, ctx)
+    local nodes = upstream_info.nodes
+    local new_nodes = {}
+    for addr, weight in pairs(nodes) do
+        local node = {}
+        local ip, port, host
+        host, port = core.utils.parse_addr(addr)
+        ip = parse_domain_for_node(host)
+        node.host = ip
+        node.port = port
+        node.weight = weight
+        table_insert(new_nodes, node)
+
+        -- Currently only supports a single upstream of the domain name.
+        -- When the upstream is `IP`, do not do any `pass_host` operation.
+        if not core.utils.parse_ipv4(host) and not core.utils.parse_ipv6(host) then
+            local pass_host = upstream_info.pass_host or "pass"
+            if pass_host == "pass" then
+                ctx.var.upstream_host = ctx.var.host
+                break
+            end
+
+            if pass_host == "rewrite" then
+                ctx.var.upstream_host = upstream_info.upstream_host
+                break
+            end
+
+            ctx.var.upstream_host = host
+            break
+        end
+    end
+    core.log.info("upstream_host: ", ctx.var.upstream_host)
+
+    local up_conf = {
+        name = upstream_info.name,
+        type = upstream_info.type,
+        nodes = new_nodes,
+        timeout = {
+            send = upstream_info.timeout and upstream_info.timeout.send or 15,
+            read = upstream_info.timeout and upstream_info.timeout.read or 15,
+            connect = upstream_info.timeout and upstream_info.timeout.connect or 15
+        }
+    }
+
+    local ok, err = upstream.check_schema(up_conf)
+    if not ok then
+        return 500, err
+    end
+
+    local matched_route = ctx.matched_route
+    upstream.set(ctx, up_conf.type .. "#route_" .. matched_route.value.id,
+                ctx.conf_version, up_conf, matched_route)
+    return
+end
+
+
+local function new_rr_obj(upstreams)
+    local server_list = {}
+    for _, upstream_obj in ipairs(upstreams) do
+        if not upstream_obj.upstream then
+            -- If the `upstream` object has only the `weight` value, it means that
+            -- the `upstream` weight value on the default `route` has been reached.
+            -- Need to set an identifier to mark the empty upstream.
+            upstream_obj.upstream = "empty_upstream"
+        end
+        server_list[upstream_obj.upstream] = upstream_obj.weight
+    end
+
+    return roundrobin:new(server_list)
+end
+
+
+function _M.access(conf, ctx)
+    if not conf or not conf.rules then
+        return
+    end
+
+    local upstreams, match_flag
+    for _, rule in pairs(conf.rules) do
+        match_flag = true
+        for _, single_match in pairs(rule.match) do
+            local expr, err = expr.new(single_match.vars)
+            if err then
+                return 500, err
+            end
+
+            match_flag = expr:eval()
+            if match_flag then
```

Comment:
Shouldn't the logic relationship between all conditions is AND rather than OR?

@membphis @moonming @spacewander What do you think?

Reply:
Yes. It is expected to use `and`.

Reply:
Yes, the expectation is `and`. Multiple matching rules can be set in `match` ,multiple conditions in `vars` are the relationship of `add`, and the relationship between multiple `vars` rules is the relationship of `or`; as long as one of the vars rules passes, it means `match` passed.

Reply:
Then the logic here should be:

```
if not expr:eval() then
    return
end
```

Reply:
We need `match_flag` to record the result of `expr:eval()`, because we need the value of `match_flag` to judge whether the `match` rule passes.

Reply:
I mean, the relationship between each rule in `match` should be `and`, so the judgement can be short circuit, we can quit the loop once one of them is `false` .